File: DeviseAlgorithm/test_2/KnapsackDP/KnapsackDP.java
/* 
    0-1背包动态规划Java实现
    时间复杂度：O(nW)
 */
 
import java.util.Scanner;
public class KnapsackDP {
    public static void main(String[] args) {
        package01_1 p = new package01_1();
        p.initPackage();
    }
}
class package01_1{
    public int max(int a,int b){
        if(a>b)
            return a;
        else
            return b;
    }
    //0-1背包动态规划
    public int[][] knapsack(int n,int C,int weight[],int value[]){
        int[][] F=new int[n+1][C+1];
        for(int i=0;i<=n;i++){
            F[i][0]=0;
        }
        for(int j=0;j<=C;j++){
            F[0][j]=0;
        }
        System.out.printf("\n编号 重量 价值\n");
        for(int i=1;i<=n;i++){
            System.out.printf("%d\t%d\t%d\n",i,weight[i-1],value[i-1]);
        }
        System.out.println("\n0—1背包动态规划表：");
        for(int i=0;i<=C;i++){
            System.out.printf("%d ",i);
        }
        System.out.println();
        for(int i=1;i<=n;i++){
            for(int j=1;j<=C;j++){
                if(weight[i-1]<=j){
                    F[i][j]=max(F[i-1][j],F[i-1][j-weight[i-1]]+value[i-1]);
                    System.out.print(F[i][j]+" ");
                }
                else {
                    F[i][j] = F[i - 1][j];
                    System.out.print(F[i][j] + " ");
                }
            }
            System.out.println();
        }
        return F;
    }
    //求最优解
    public void Judge(int n,int C,int weight[],int F[][]){
        int state[]=new int[n];
        int j=C;
        for(int i=n-1;i>=1;i--){
            if(F[i][j]>F[i-1][j]){
                state[i]=1;
                j=j-weight[i-1];
            }
            else
                state[i]=0;
        }
        System.out.println("选中的物品编号是:");
        for(int i=0;i<n;i++) {
            if (state[i] == 1)
                System.out.print(i+" ");
        }
       //System.out.println(" ");
    }
    //随机生成物品重量
    public void RandomWeight(int weight[], int len, int max) {
        for (int i = 0; i < weight.length; i++) {
            weight[i] = (int) (Math.random() * max + 1);
        }
        System.out.println("物品的随机重量：");
        for (int i = 0; i < weight.length; i++) {
            System.out.print(weight[i] + " ");
        }
    }
    //随机生成物品价值
    public void RandomValue(int value[], int len, int max) {
        for (int i = 0; i < value.length; i++) {
            value[i] = (int) (Math.random() * max + 1);
        }
        System.out.println("物品的随机价值：");
        for (int i = 0; i < value.length; i++) {
            System.out.print(value[i] + " ");
        }
    }
    //初始化
    public void initPackage(){
        try (Scanner sc = new Scanner(System.in)) {
            System.out.print("请输入物品个数：");
            int N = sc.nextInt();
            System.out.print("请输入背包的最大容量：");
            int C = sc.nextInt();
            int[] weight = new int[N];
            int[] value = new int[N];
            RandomWeight(weight,N,100);
            System.out.println();
            RandomValue(value,N,100);
            //int F[][] = knapsack(N,C,weight,value);
            long startTime = System.currentTimeMillis();//起始时间
            int F[][] = knapsack(N,C,weight,value);
            Judge(N,C,weight,F);

            System.out.println("\n最大价值为："+F[N][C]);
            long endTime = System.currentTimeMillis();//结束时间
            System.out.println("\n程序运行时间： " + (endTime - startTime ) + "ms");
        }
    }
}

File: DeviseAlgorithm/test_2/KnapsackDP/KnapsackDP.py
# 0-1背包动态规划Python实现
# 时间复杂度;O(n*c)
import random
import time
# 求解背包
def bag(n, c, w, v):
    # 置零，表示初始状态
    value = [[0 for j in range(c + 1)] for i in range(n + 1)]
    for i in range(1, n + 1):
        for j in range(1, c + 1):
            value[i][j] = value[i - 1][j]
            # 背包总容量够放当前物体，遍历前一个状态考虑是否置换
            if j >= w[i - 1] and value[i][j] < value[i - 1][j - w[i - 1]] + v[i - 1]:
                value[i][j] = value[i - 1][j - w[i - 1]] + v[i - 1]
    for x in value:
        print(x)
    return value
# 最大值和最优解
def PackageMax(n, c, w, value):
    print('最大价值为:', value[n][c])
    x = [False for i in range(n)]
    j = c
    for i in range(n, 0, -1):
        if value[i][j] > value[i - 1][j]:
            x[i - 1] = True
            j -= w[i - 1]
    print("背包中所装物品为:")
    for i in range(n):
        if x[i]:
            print('第', i+1, '个,', end='')

# 随机n个物品的重量
def RandomWeight(n):
    arr = []
    for i in range(n):
        arr.append(random.randint(1, 100))
    return arr

# 随机n个物品的价值
def RandomValue(n):
    list = []
    for i in range(n):
        list.append(random.randint(1, 100))
    return list

if __name__ == '__main__':
    n= (int)(input("物品数量："))
    c = (int)(input("背包最大容量："))
    
    w=RandomWeight(n)
    v=RandomValue(n)
    print("物品的重量:")
    print(w)
    print("物品的价值：")
    print(v)
    start_time=time.time()
    value = bag(n, c, w, v)
    PackageMax(n, c, w, value)
    end_time=time.time()
    print("\n运行时间：",float(end_time-start_time)*1000.0,"ms")



